File: leetcode_problems/cpp/lc110_balanced_binary_tree.cpp
#include <algorithm>

struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution
{
private:
    bool balanced;
    int height(TreeNode *node)
    {
        if (node == nullptr)
        {
            return 0;
        }
        int left = height(node->left);
        int right = height(node->right);
        if (std::abs(left - right) > 1)
        {
            balanced = false;
        }
        return std::max(left, right) + 1;
    }

public:
    bool
    isBalanced(TreeNode *root)
    {
        balanced = true;
        height(root);
        return balanced;
    }
};

File: leetcode_problems/typescript/src/lc110_balanced_binary_tree.ts
function isBalancedHeight(
  root: TreeNode | null,
  balanced: { value: boolean }
): number {
  if (root === null) {
    return 0;
  }

  let left = isBalancedHeight(root.left, balanced);
  let right = isBalancedHeight(root.right, balanced);
  if (Math.abs(left - right) > 1) {
    balanced.value = false;
  }
  return Math.max(left, right) + 1;
}

function isBalanced(root: TreeNode | null): boolean {
  let balanced = { value: true };
  isBalancedHeight(root, balanced);
  return balanced.value;
}
